src/deluge/gui/views/automation_instrument_clip_view.cpp

```diff
@@ -0,0 +1,3138 @@
+	if (copiedParamAutomation.nodes) {
+		GeneralMemoryAllocator::get().dealloc(copiedParamAutomation.nodes);
+		copiedParamAutomation.nodes = NULL;
+		copiedParamAutomation.numNodes = 0;
+	}
+
+	int32_t startPos = getPosFromSquare(0);
+	int32_t endPos = getPosFromSquare(kDisplayWidth);
+	if (startPos == endPos) {
+		return;
+	}
+
+	InstrumentClip* clip = getCurrentClip();
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+
+	ModelStackWithTimelineCounter* modelStack = currentSong->setupModelStackWithCurrentClip(modelStackMemory);
+
+	ModelStackWithAutoParam* modelStackWithParam =
+	    getModelStackWithParam(modelStack, clip, clip->lastSelectedParamID, clip->lastSelectedParamType);
+
+	if (modelStackWithParam && modelStackWithParam->autoParam) {
+
+		bool isPatchCable = (modelStackWithParam->paramCollection
+		                     == modelStackWithParam->paramManager->getPatchCableSetAllowJibberish());
+		// Ok this is cursed, but will work fine so long as
+		// the possibly invalid memory here doesn't accidentally
+		// equal modelStack->paramCollection.
+
+		modelStackWithParam->autoParam->copy(startPos, endPos, &copiedParamAutomation, isPatchCable,
+		                                     modelStackWithParam);
+
+		if (copiedParamAutomation.nodes) {
+			numericDriver.displayPopup(HAVE_OLED ? "Automation copied" : "COPY");
+			setDisplayParameterNameTimer();
+			return;
+		}
+	}
+
+	numericDriver.displayPopup(HAVE_OLED ? "No automation to copy" : "NONE");
+	setDisplayParameterNameTimer();
+}
+
+void AutomationInstrumentClipView::pasteAutomation() {
+	if (!copiedParamAutomation.nodes) {
+		numericDriver.displayPopup(HAVE_OLED ? "No automation to paste" : "NONE");
+		setDisplayParameterNameTimer();
+		return;
+	}
+
+	int32_t startPos = getPosFromSquare(0);
+	int32_t endPos = getPosFromSquare(kDisplayWidth);
+
+	int32_t pastedAutomationWidth = endPos - startPos;
+	if (pastedAutomationWidth == 0) {
+		return;
+	}
+
+	float scaleFactor = (float)pastedAutomationWidth / copiedParamAutomation.width;
+
+	InstrumentClip* clip = getCurrentClip();
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+
+	ModelStackWithTimelineCounter* modelStack = currentSong->setupModelStackWithCurrentClip(modelStackMemory);
+
+	ModelStackWithAutoParam* modelStackWithParam =
+	    getModelStackWithParam(modelStack, clip, clip->lastSelectedParamID, clip->lastSelectedParamType);
+
+	if (modelStackWithParam && modelStackWithParam->autoParam) {
+		Action* action = actionLogger.getNewAction(ACTION_AUTOMATION_PASTE, false);
+
+		if (action) {
+			action->recordParamChangeIfNotAlreadySnapshotted(modelStackWithParam, false);
+		}
+
+		bool isPatchCable = (modelStackWithParam->paramCollection
+		                     == modelStackWithParam->paramManager->getPatchCableSetAllowJibberish());
+		// Ok this is cursed, but will work fine so long as
+		// the possibly invalid memory here doesn't accidentally
+		// equal modelStack->paramCollection.
```

Comment:
I'm not sure I understand this comment but I don't think I like it :sweat_smile:

Reply:
this was Rohan's comment hahah. I don't understand it either :S.

Reply:
I'm not it even applies to this code anymore now that I look at it again. If we get weirdness with params again this is probably a good place to go hunting though...